src/python/pants/build_graph/address.py

```diff
@@ -244,18 +244,21 @@ class BuildFileAddress(Address):
   :API: public
   """
 
-  def __init__(self, build_file, target_name=None):
+  def __init__(self, build_file=None, target_name=None, rel_path=None):
     """
     :param build_file: The build file that contains the object this address points to.
     :type build_file: :class:`pants.base.build_file.BuildFile`
+    :param string rel_path: The path relative to root_dir where the BUILD file is located.
     :param string target_name: The name of the target within the BUILD file; defaults to the default
                                target, aka the name of the BUILD file parent dir.
 
     :API: public
     """
-    spec_path = os.path.dirname(build_file.relpath)
+    rel_path = rel_path or build_file.relpath
```

Comment:
Do you think we ought to have a tailored exception here if both rel_path and build_file are None?
Also, I think this will fail if `rel_path` is "" and build_file is None.